### PR TITLE
Consider existing deployments correctly when creating a new deployment

### PR DESCRIPTION
**Pull request by abhgupta**
Closes https://github.com/openshift/origin/issues/2932
Closes https://github.com/openshift/origin/issues/2447

We are now correctly checking for the latest deployment to decide whether to create a new deployment or not. Also, added test cases that validate this scenario.

### CODE REVIEW

**Comment by abhgupta** [comment]
@derekwaynecarr @ironcladlou

**Comment by abhgupta** [comment]
@danmcp @smarterclayton Please review.

**Comment by derekwaynecarr** [comment]
That looks like it would do the trick and stop multiple attempts to create what was already created.  In a multi controller scenario, should there also be a corresponding catch to the code that ignores already exists errors after the create and does log or produce an error when that happens?  I was but by similar things (but on the delete case)

**Comment by smarterclayton** [comment]
Yes, we should ignore already exists errors in most of these cases

File: pkg/deploy/controller/deploymentconfig/controller.go [hunk 1]
```diff
@@ -135,7 +142,6 @@ func (c *DeploymentConfigController) Handle(config *deployapi.DeploymentConfig)
 		// If the deployment was already created, just move on. The cache could be stale, or another
 		// process could have already handled this update.
 		if errors.IsAlreadyExists(err) {
-			c.recorder.Eventf(config, "alreadyExists", "Deployment already exists for DeploymentConfig: %s", deployutil.LabelForDeploymentConfig(config))
 			glog.V(4).Infof("Deployment already exists for DeploymentConfig %s", deployutil.LabelForDeploymentConfig(config))
```
Comment (abhgupta) [inline pkg/deploy/controller/deploymentconfig/controller.go]: removed the event and retained the log statement for the case when the deployment already exists - presumably from another controller running simultaneously.

**Comment by abhgupta** [comment]
@smarterclayton @derekwaynecarr PR updated.

**Comment by derekwaynecarr** [comment]
Lgtm

**Comment by smarterclayton** [comment]
Approved [merge]

File: pkg/deploy/controller/deploymentconfig/controller.go [hunk 2]
```diff
@@ -62,7 +62,14 @@ func (c *DeploymentConfigController) Handle(config *deployapi.DeploymentConfig)
 		return fmt.Errorf("couldn't list Deployments for DeploymentConfig %s: %v", deployutil.LabelForDeploymentConfig(config), err)
 	}
 	var inflightDeployment *kapi.ReplicationController
+	latestDeploymentExists := false
 	for _, deployment := range existingDeployments.Items {
+		// check if this is the latest deployment
+		// we'll return after we've dealt with the multiple-active-deployments case
+		if deployutil.DeploymentVersionFor(&deployment) == config.LatestVersion {
+			latestDeploymentExists = true
```
Comment (ironcladlou) [inline pkg/deploy/controller/deploymentconfig/controller.go]: I'm late to the party: why not just return here?

Comment (abhgupta) [inline pkg/deploy/controller/deploymentconfig/controller.go]:
```
// we'll return after we've dealt with the multiple-active-deployments case
```
Did try to cover that with a comment - we would like to identify and clean up concurrently active deployments

**Pull request merged**